### ISSUE TITLE
The `:checked` proposal

### ISSUE THREAD

**Issue by thoughtstream** [Raku/RakuDoc-GAMMA#19]
@zag's submission included a mechanism for check-marking any block
via a `:checked` meta-option, as follows:

> ### `:checked`
>
> This attribute indicates that a checkbox should be added to that block.
> It possible to apply a `:checked` attributes to any block.
> The most common use of this option is to create task lists.
>
> For unchecked checkbox the `:!checked` is used.
>
> The task list item marker (checkbox) is added to the output.
> In an HTML output, this would be represented as an
> `<input type="checkbox">` element.
>
> The specification does not define how these checkboxes are interacted with.
> Implementors are free to choose whether they render them as disabled,
> unchangeable elements, or handle dynamic interactions like checking and unchecking
> in the final rendered document.

I agree with @finanalyst that, because we consider RakuDoc to be
a static, non-interactive document-description language,
actual checkboxes are not appropriate as part of the core feature set.

On the other hand, I can certainly see the value of being able to specify
a (static, non-interactive) list of checked and unchecked items.
For example:

> _The project originally consisted of five phases, of which
> two are already complete and two have been abandoned:_
>
> ☑︎ _Investigate existing solutions_
> ☑︎ _Define a minimal initial feature set_
> ☐ _Implement this minimal set of features_
> ☒ _Secure 100 million in venture capital_
> ☒ _Abscond to the Bahamas with the cash_

But these are not actually checkboxes; they are custom ***bullets***.

Which RakuDoc does not currently support, as we state explicitly:

> The bulleting strategy used for different levels within a nested list
> is entirely up to the renderer.

But perhaps we should consider supporting custom bullets?

Perhaps something like:
```rakudoc
=for item :bullet< E<BALLOT BOX WITH CHECK> >
Investigate existing solutions

=for item :bullet< E<BALLOT BOX WITH CHECK> >
Define a minimal initial feature set

=for item :bullet< E<BALLOT BOX> >
Implement this minimal set of features

=for item :bullet< E<BALLOT BOX WITH CROSS> >
Secure 100 million in venture capital

=for item :bullet< E<BALLOT BOX WITH CROSS> >
Abscond to the Bahamas with the cash
```

That would then also allow users to choose other bulleting strategies,
such as:
```rakudoc
=for item :bullet< E<tick symbol> >
Investigate existing solutions

=for item :bullet< E<tick symbol> >
Define a minimal initial feature set

=for item :bullet< E<anticlockwise arrows> >
Implement this minimal set of features

=for item :bullet< E<prohibited> >
Secure 100 million in venture capital

=for item :bullet< E<prohibited> >
Abscond to the Bahamas with the cash
```
...to produce:
> ✅ _Investigate existing solutions_
> ✅ _Define a minimal initial feature set_
> 🔄 _Implement this minimal set of features_
> 🚫 _Secure 100 million in venture capital_
> 🚫 _Abscond to the Bahamas with the cash_

Item bullets could also be preconfigured for an entire list:
```rakudoc
=begin section

=config item :bullet< E<globe showing Europe and Africa> >

The major sources of sustainable energy are:
    =item wind
    =item hydroelectric
    =item solar
    =item geothermal
    =item fusion (eventually)
=end section
```
...which would produce:
> The major sources of sustainable energy are:
> 🌍 wind
> 🌍 hydroelectric
> 🌍 solar
> 🌍 geothermal
> 🌍 fusion (eventually)

I can imagine other uses as well, such as localized notes (for those who prefer more control
than the `N<>` instruction provides):
```rakudoc
Paragraphs are useful because humans can only focus on a few pieces of information at once.⁽¹⁾
Paragraphs are one way of aggregating small amounts of related information, so that the
resulting “chunk” can fit into a single slot of the reader’s limited short-term memory.
Paragraphs enable the physical structure of a piece of writing to reflect and emphasize
its logical structure. Adding comments at the start of each paragraph further enhances
the chunking by explicitly summarizing the purpose⁽²⁾ of each chunk.

━━━━━━━━━━━━
=for item :bullet<⁽¹⁾>
An idea made famous in 1956 by George A. Miller in I<The Magical Number Seven, Plus or Minus Two>
(The Psychological Review, 1956, vol. 63, pp. 81-97).

=for item :bullet<⁽²⁾>
The B<purpose>, not the B<actions>. Paragraph comments need to explain I<why> the code is needed,
not merely paraphrase what it’s doing.
```
And, yes, the above example does indeed imply that a `:bullet` can be specified as two-or-more characters,
not just as a single character.

Of course, a renderer should be still permitted to ignore or override
a `:bullet` option (e.g. if the renderer is ASCII-only or non-visual),
so we would need to characterize `:bullet` as being a ***request*** 
for the specified bullet.

On the other hand, maybe we don’t need user-configurable bullets at all
...or don't need them at this time. I admit that I find myself going back and forth 
in my own mind on the idea. Mostly I tend to like it more than I loathe it, 
but I’d be happier if we collectively discussed its pros and cons a little further.

**Comment by finanalyst** [comment]
@thoughtstream As I thought about `:checked` and before reading this issue, I came to the same conclusion that a static indicator was something that could be used. The idea of designating it as a custom bullet is brilliant and the extension to other customisations is actually something I have wanted too. @thoughtstream  I am constantly amazed how you are able to expand a concept to increase usability.

So, I am definitely in favour of this proposal.

I think that we can say that `=itemN` has an implied `:bullet` for each value of N. A RakuDoc renderer should provide a default for at least `N=1..4`. 

Obviously,
```
=config item1 :bullet< E<1234>>
=config item2 :bullet< E<1256>>
```
would be standard RakuDoc to provide a whole bulleting strategy.

I have faced some problems getting Unicode entities to work, so I think that we should stipulate for `:bullet` that if a renderer is unable to use an author-supplied bullet for an item, the fall back is the default rather than an error glyph, and add an error message as described in the standard.

Regarding `:bullet` for blocks other than `=item`, I'm unconvinced. Notations and numbered lists and headings adhere to a numbering system. We have already noted in the standard how numbering systems should be rendered by default. The actual numbering system, eg using Latin or Chinese numerals, can be defined in a renderer specific way, for example to leverage the standard HTML ordered list attributes.

**Comment by thoughtstream** [comment]
> thoughtstream I am constantly amazed how you are able to expand a concept to increase usability.

It's kind of you to say that. If I do have that ability, it’s mostly because I spent over a decade-and-a-half 
learning at the feet of a [grand master](https://en.wikipedia.org/wiki/Larry_Wall) of the technique. We spent most of those 15 years 
wading though hundreds of special-case feature requests and extracting highly generalized
and orthogonal Raku solutions from them. :-)

> I think that we can say that `=itemN` has an implied `:bullet` for each value of N. A RakuDoc
> renderer should provide a default for at least `N=1..4`.

Agreed. Though we cannot specify what those defaults are, as they are likely
to be different depending on the target format of each renderer.

> Obviously,
> ```
> =config item1 :bullet< E<1234>>
> =config item2 :bullet< E<1256>>
> ```
> would be standard RakuDoc to provide a whole bulleting strategy.

Indeed. Though I’m not sure we should encourage anyone to use "Ӓ" and "Ө" as bullets ;-)


> I have faced some problems getting Unicode entities to work, so I think that we should
> stipulate for `:bullet` that if a renderer is unable to use an author-supplied bullet for an
> item, the fall back is the default rather than an error glyph, and add an error message as
> described in the standard.

Agreed.


> Regarding `:bullet` for blocks other than `=item`, I'm unconvinced.

Likewise. Especially as there are already plausible ways to achieve
(for example) checkmarked headings, checkmarked code blocks, 
or checkmarked plain paragraphs:

```rakudoc
=begin item :bullet< E<BALLOT BOX WITH CHECK> >
=head3 Stage One
=end item

=begin item :bullet< E<BALLOT BOX WITH CROSS> >
=begin code
# This code is complete and has been deployed
$x = $x + 1;
=end code
=end item

=for item :bullet< E<BALLOT BOX> >
This text is still being reviewed.
```
...to produce:

### ☑︎ Stage One

`  ☒  # This code is complete and has been deployed    `
`     $x = $x + 1;                                     `

☐ This text is still being reviewed.

In other words, if a block needs a bullet, then that block is intrinsically a list item,
and should be specified as such.

**Comment by finanalyst** [comment]
closing as implemented
